src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -312,14 +312,19 @@ private TCommandResult ProcessResponse(ConnectionId connectionId, CommandMessage
                     if (materializedDocument.TryGetValue("errmsg", out errmsgBsonValue) && errmsgBsonValue.IsString)
                     {
                         var errmsg = errmsgBsonValue.ToString();
+                        message = string.Format("Command {0} failed: {1}.", commandName, errmsg);
+                        // https://jira.mongodb.org/browse/CSHARP-2678
                         if (materializedDocument.TryGetValue("code", out var errorCode)
-                            && errorCode.AsInt32 == 20
+                            && errorCode.ToInt32() == 20
                             && errmsg.StartsWith("Transaction numbers"))
                         {
-                            errmsg = "This MongoDB deployment does not support retryable writes. "
-                                     + "Please add retryWrites=false to your connection string.";
+                            const string friendlyErrorMessage =
+                                "This MongoDB deployment does not support retryable writes. " +
+                                "Please add retryWrites=false to your connection string.";
+                            var innerException =
+                                new MongoCommandException(connectionId, message, _command, materializedDocument);
+                            throw new MongoException(friendlyErrorMessage, innerException);
```

Comment:
I would leave all lines before 334 untouched (i.e. continue to create the original exception exactly as before), and replace line 334 with something like:

    var exception = new MongoCommandException(connectionId, message, _command, materializedDocument);
    if (IsRetryableWriteExceptionAndDeploymentDoesNotSupportRetryableWrites(exception))
    {
        throw WrapNotSupportedRetryableWriteExcepction(exception);
    }
    else
    {
        throw exception;
    }

pushing some of the logic into helper functions and making it real clear that we are either throwing the original exception or wrapping it.

Reply:
Or if you want to be more general:

    var exception = ...
    exception = WrapIfNecessary(exception);
    throw exception;

Reply:
Updated!